src/hdf5_back_gen.py

```diff
@@ -0,0 +1,496 @@
+{t.cpptype} x;
+for (unsigned int k = 0; k < jlen; ++k) {{
+    x[VLRead<{t.sub[1].cpptype}, {t.sub[1].dbtype}>(buf + offset + itemsize*k)] = *reinterpret_cast<{t.sub[2].cpptype}*>(buf + offset + itemsize*k + CYCLUS_SHA1_SIZE);
+}}
+{NO_CLOSE}\
+"""
+
+MAP_STRING_STRING_READER = """\
+hid_t field_type = H5Tget_member_type(tb_type, j);
+hid_t item_type = H5Tget_super(field_type);
+hid_t key_type = H5Tget_member_type(item_type, 0);
+size_t nullpos;
+hsize_t fieldlen;
+H5Tget_array_dims2(field_type, &fieldlen);
+unsigned int itemsize = col_sizes_[table][j] / fieldlen;
+unsigned int keylen = H5Tget_size(key_type);
+unsigned int vallen = itemsize - keylen;
+{t.cpptype} x;
+for (unsigned int k = 0; k < fieldlen; ++k) {{
+    {t.sub[1].cpptype} key = {t.sub[1].cpptype}(buf + offset + itemsize*k, keylen);
+    nullpos = key.find('\0');
+    if (nullpos != {t.sub[1].cpptype}::npos)
+        key.resize(nullpos);
+    {t.sub[2].cpptype} val = {t.sub[2].cpptype}(buf + offset + itemsize*k + keylen, vallen);
+    nullpos = val.find('\0');
+    if (nullpos != {t.sub[2].cpptype}::npos)
+        val.resize(nullpos);
+    x[key] = val;
+}}
+{H5TCLOSE_MULTI}\
+"""
+
+MAP_STRING_VL_STRING_READER = """\
+hid_t field_type = H5Tget_member_type(tb_type, j);
+size_t nullpos;
+hsize_t fieldlen;
+H5Tget_array_dims2(field_type, &fieldlen);
+unsigned int itemsize = col_sizes_[table][j] / fieldlen;
+unsigned int keylen = itemsize - CYCLUS_SHA1_SIZE;
+{t.cpptype} x;
+for (unsigned int k = 0; k < fieldlen; ++k) {{
+    {t.sub[1].cpptype} key = {t.sub[1].cpptype}(buf + offset + itemsize*k, keylen);
+    nullpos = key.find('\0');
+    if (nullpos != {t.sub[1].cpptype}::npos)
+        key.resize(nullpos);
+    x[key] = VLRead<{t.sub[2].cpptype}, {t.sub[2].dbtype}>(buf + offset + itemsize*k + keylen);
+}}
+{H5TCLOSE}\
+"""
+
+MAP_VL_STRING_STRING_READER = """\
+hid_t field_type = H5Tget_member_type(tb_type, j);
+size_t nullpos;
+hsize_t fieldlen;
+H5Tget_array_dims2(field_type, &fieldlen);
+unsigned int itemsize = col_sizes_[table][j] / fieldlen;
+unsigned int vallen = itemsize - CYCLUS_SHA1_SIZE;
+{t.cpptype} x;
+for (unsigned int k = 0; k < fieldlen; ++k) {{
+    {t.sub[2].cpptype} val = {t.sub[2].cpptype}(buf + offset + itemsize*k + CYCLUS_SHA1_SIZE, vallen);
+    nullpos = val.find('\0');
+    if (nullpos != {t.sub[2].cpptype}::npos)
+        val.resize(nullpos);
+    x[VLRead<{t.sub[1].cpptype}, {t.sub[1].dbtype}>(buf + offset + itemsize*k)] = val;
+}}
+{H5TCLOSE}\
+"""
+
+MAP_VL_STRING_VL_STRING_READER = """\
+hid_t field_type = H5Tget_member_type(tb_type, j);
+size_t nullpos;
+hsize_t fieldlen;
+H5Tget_array_dims2(field_type, &fieldlen);
+unsigned int itemsize = 2*CYCLUS_SHA1_SIZE;
+{t.cpptype} x;
+for(unsigned int k = 0; k < fieldlen; ++k) {{
+    x[VLRead<{t.sub[1].cpptype}, {t.sub[1].dbtype}>(buf + offset + itemsize*k)] = VLRead<{t.sub[2].cpptype}, {t.sub[2].dbtype}>(buf + offset + itemsize*k + CYCLUS_SHA1_SIZE);
+}}
+{H5TCLOSE}\
+"""
```

Comment:
Rather than doing the trailing `\` on the first and last line of all of these, you can instead end the quote with `""".strip()` rather than just `"""`